PowerEditor/src/MISC/Common/Common.cpp

```diff
@@ -1764,60 +1764,105 @@ bool Version::isCompatibleTo(const Version& from, const Version& to) const
 	return false;
 }
 
-
-/*
-bool doesFileExist(const wchar_t* filePath)
-{
-	DWORD dwAttrib = ::GetFileAttributesW(filePath);
-	return (dwAttrib != INVALID_FILE_ATTRIBUTES && !(dwAttrib & FILE_ATTRIBUTE_DIRECTORY));
-}
-*/
-
 struct GetAttrParamResult {
 	std::wstring _filePath;
-	DWORD _fileAttr;
+	DWORD _fileAttr = INVALID_FILE_ATTRIBUTES;
+	bool _isNetworkFailure = true;
 };
 
 DWORD WINAPI getFileAttributesWorker(void* data)
 {
 	GetAttrParamResult* inAndOut = static_cast<GetAttrParamResult*>(data);
 	inAndOut->_fileAttr = ::GetFileAttributesW(inAndOut->_filePath.c_str());
+	inAndOut->_isNetworkFailure = false;
 	return TRUE;
 };
 
-DWORD getFileAttrWaitFewSec(const wchar_t* filePath)
+DWORD getFileAttrWaitFewSec(const wchar_t* filePath, DWORD milleSec2wait, bool* isNetWorkProblem)
 {
-	static GetAttrParamResult data;
+	GetAttrParamResult data;
 	data._fileAttr = INVALID_FILE_ATTRIBUTES;
 	data._filePath = filePath;
+	data._isNetworkFailure = true;
+
 	HANDLE hThread = ::CreateThread(NULL, 0, getFileAttributesWorker, &data, 0, NULL);
 	if (!hThread)
 	{
 		return false;
 	}
 
-	// Wait for 3 sec
-	::WaitForSingleObject(hThread, 3000);
+	// Wait for few sec
+	::WaitForSingleObject(hThread, milleSec2wait == 0 ? 1000 : milleSec2wait);
+	TerminateThread(hThread, 2);

```

Comment:
For threads, the TerminateThread is a really nasty shutdown, so if possible it should not be a part of the usual code flow.

So I would use something like (you said the `CancelSynchronousIo` WINAPI does not work for you - in that case, delete my WAIT_TIMEOUT part entirely):

// wait for our worker thread to complete or cancel it when the required timeout has elapsed
DWORD dwWaitStatus = ::WaitForSingleObject(hThread, milleSec2wait == 0 ? 1000 : milleSec2wait);
switch (dwWaitStatus)
{
	case WAIT_OBJECT_0:
		// Ok, the state of our worker thread is signaled, so it finished itself in the timeout given
		// - nothing else to do here, except the thread handle closing later
		break;

	case WAIT_TIMEOUT:
		// the timeout interval elapsed, but the worker's state is still non-signaled
		// - try to cancel our worker thread's synchronous IO from here
		if (!::CancelSynchronousIo(hThread))
		{
			// if the CancelSynchronousIo cannot find a request to cancel,
			// its return value is 0 and the GetLastError() returns ERROR_NOT_FOUND
			dwRet = ::GetLastError();
			if (dwRet != ERROR_NOT_FOUND)
				::TerminateThread(hThread, WAIT_TIMEOUT);
		}
		break;

		// any other dwWaitStatus is a BAD one here
	default:
		// WAIT_FAILED or WAIT_ABANDONED
		::TerminateThread(hThread, dwWaitStatus);
		break;
}

Reply:
@donho 
> To reproduce the delay, you just need to open 1 network file then cut off the connection, open some new tab, make it dirty, and close the dirty tab. When the focus come back to the disconnected network file, you'll see the huge delay.

Thank you, with this STR I am finally able to reproduce the issue steadily (my previous attempts mostly ended up with immediate "Keep non existing file" dlg).

I will recheck the `CancelSynchronousIo` sln against your latest commits and let you know.